File: src/core/formula.rs
//! Formula data structures from Homebrew API

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

use super::bottle::BottleSpec;

/// A Homebrew formula (package definition)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Formula {
    /// Package name (e.g., "jq")
    pub name: String,

    /// Full name including tap (e.g., "homebrew/core/jq")
    pub full_name: String,

    /// Tap name (e.g., "homebrew/core")
    #[serde(default)]
    pub tap: String,

    /// Package description
    pub desc: Option<String>,

    /// Homepage URL
    pub homepage: Option<String>,

    /// License
    pub license: Option<String>,

    /// Version information
    pub versions: Versions,

    /// Bottle (precompiled binary) specification
    #[serde(default)]
    pub bottle: BottleSpec,

    /// Runtime dependencies
    #[serde(default)]
    pub dependencies: Vec<String>,

    /// Build-time dependencies
    #[serde(default)]
    pub build_dependencies: Vec<String>,

    /// Optional dependencies
    #[serde(default)]
    pub optional_dependencies: Vec<String>,

    /// Test dependencies
    #[serde(default)]
    pub test_dependencies: Vec<String>,

    /// Recommended dependencies
    #[serde(default)]
    pub recommended_dependencies: Vec<String>,

    /// Whether this package is keg-only (not linked to PATH by default)
    #[serde(default)]
    pub keg_only: bool,

    /// Reason for being keg-only
    pub keg_only_reason: Option<KegOnlyReason>,

    /// Whether this package is deprecated
    #[serde(default)]
    pub deprecated: bool,

    /// Deprecation date
    pub deprecation_date: Option<String>,

    /// Deprecation reason
    pub deprecation_reason: Option<String>,

    /// Whether this package is disabled
    #[serde(default)]
    pub disabled: bool,

    /// Disable date
    pub disable_date: Option<String>,

    /// Disable reason
    pub disable_reason: Option<String>,

    /// Post-install caveats
    pub caveats: Option<String>,

    /// URLs for various versions
    #[serde(default)]
    pub urls: HashMap<String, UrlSpec>,

    /// Revision number
    #[serde(default)]
    pub revision: u32,

    /// Version scheme
    #[serde(default)]
    pub version_scheme: u32,

    /// Link overwrite files
    #[serde(default)]
    pub link_overwrite: Vec<String>,

    /// Post-install defined
    #[serde(default)]
    pub post_install_defined: bool,

    /// Service defined
    #[serde(default)]
    pub service: Option<serde_json::Value>,

    /// Analytics install data (30 days)
    pub analytics: Option<Analytics>,

    /// Installation count (30 days)
    #[serde(default)]
    pub analytics_install_on_request_30d: Option<i64>,
}

/// Version information
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Versions {
    /// Stable version
    pub stable: String,

    /// HEAD version (from git)
    pub head: Option<String>,

    /// Bottle available
    #[serde(default)]
    pub bottle: bool,
}

/// URL specification for sources
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UrlSpec {
    pub url: String,
    #[serde(default)]
    pub tag: Option<String>,
    #[serde(default)]
    pub revision: Option<String>,
    #[serde(default)]
    pub using: Option<String>,
    #[serde(default)]
    pub checksum: Option<String>,
}

/// Reason for being keg-only
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KegOnlyReason {
    pub reason: String,
    pub explanation: Option<String>,
}

/// Analytics data
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Analytics {
    #[serde(default)]
    pub install: AnalyticsInstall,
    #[serde(default)]
    pub install_on_request: AnalyticsInstall,
    #[serde(default)]
    pub build_error: AnalyticsBuildError,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct AnalyticsInstall {
    #[serde(rename = "30d", default)]
    pub thirty_days: HashMap<String, i64>,
    #[serde(rename = "90d", default)]
    pub ninety_days: HashMap<String, i64>,
    #[serde(rename = "365d", default)]
    pub one_year: HashMap<String, i64>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct AnalyticsBuildError {
    #[serde(rename = "30d", default)]
    pub thirty_days: HashMap<String, i64>,
}

impl Formula {
    /// Get the latest stable version string
    pub fn version(&self) -> &str {
        &self.versions.stable
    }

    /// Get the stable version including revision suffix, if any
    pub fn version_with_revision(&self) -> String {
        if self.revision > 0 {
            format!("{}_{}", self.versions.stable, self.revision)
        } else {
            self.versions.stable.clone()
        }
    }

    /// Check if this formula has a bottle for the given platform tag
    pub fn has_bottle(&self, tag: &str) -> bool {
        self.bottle
            .stable
            .as_ref()
            .map(|files| files.files.contains_key(tag))
            .unwrap_or(false)
    }

    /// Get bottle files for a specific tag
    pub fn bottle_for_tag(&self, tag: &str) -> Option<&super::bottle::BottleFile> {
        self.bottle
            .stable
            .as_ref()
            .and_then(|files| files.files.get(tag))
    }

    /// Get all available bottle tags
    pub fn available_bottle_tags(&self) -> Vec<&str> {
        self.bottle
            .stable
            .as_ref()
            .map(|files| files.files.keys().map(|s| s.as_str()).collect())
            .unwrap_or_default()
    }

    /// Get all runtime dependencies
    pub fn all_dependencies(&self) -> Vec<&str> {
        self.dependencies.iter().map(|s| s.as_str()).collect()
    }

    /// Check if this is a simple formula (no post-install hooks, no services)
    pub fn is_simple(&self) -> bool {
        !self.post_install_defined && self.service.is_none()
    }

    /// Get formatted display name with version
    pub fn display_name(&self) -> String {
        format!("{} {}", self.name, self.versions.stable)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_formula_version() {
        let formula = Formula {
            name: "jq".to_string(),
            full_name: "homebrew/core/jq".to_string(),
            tap: "homebrew/core".to_string(),
            desc: Some("Lightweight JSON processor".to_string()),
            homepage: Some("https://jqlang.github.io/jq/".to_string()),
            license: Some("MIT".to_string()),
            versions: Versions {
                stable: "1.7.1".to_string(),
                head: None,
                bottle: true,
            },
            bottle: BottleSpec::default(),
            dependencies: vec![],
            build_dependencies: vec![],
            optional_dependencies: vec![],
            test_dependencies: vec![],
            recommended_dependencies: vec![],
            keg_only: false,
            keg_only_reason: None,
            deprecated: false,
            deprecation_date: None,
            deprecation_reason: None,
            disabled: false,
            disable_date: None,
            disable_reason: None,
            caveats: None,
            urls: HashMap::new(),
            revision: 0,
            version_scheme: 0,
            link_overwrite: vec![],
            post_install_defined: false,
            service: None,
            analytics: None,
            analytics_install_on_request_30d: None,
        };

        assert_eq!(formula.version(), "1.7.1");
        assert_eq!(formula.display_name(), "jq 1.7.1");
        assert!(formula.is_simple());
    }
}

File: swift/Sources/ColdbrewKit/Registry/Formula.swift
import Foundation

public struct Formula: Codable, Equatable, Sendable {
    public var name: String
    public var fullName: String
    public var tap: String
    public var desc: String?
    public var homepage: String?
    public var license: String?
    public var versions: FormulaVersions
    public var bottle: BottleSpec
    public var dependencies: [String]
    public var buildDependencies: [String]
    public var optionalDependencies: [String]
    public var testDependencies: [String]
    public var recommendedDependencies: [String]
    public var kegOnly: Bool
    public var revision: UInt32
    public var versionScheme: UInt32

    enum CodingKeys: String, CodingKey {
        case name
        case fullName = "full_name"
        case tap
        case desc
        case homepage
        case license
        case versions
        case bottle
        case dependencies
        case buildDependencies = "build_dependencies"
        case optionalDependencies = "optional_dependencies"
        case testDependencies = "test_dependencies"
        case recommendedDependencies = "recommended_dependencies"
        case kegOnly = "keg_only"
        case revision
        case versionScheme = "version_scheme"
    }

    public init(
        name: String,
        fullName: String,
        tap: String = "homebrew/core",
        desc: String? = nil,
        homepage: String? = nil,
        license: String? = nil,
        versions: FormulaVersions,
        bottle: BottleSpec = BottleSpec(),
        dependencies: [String] = [],
        buildDependencies: [String] = [],
        optionalDependencies: [String] = [],
        testDependencies: [String] = [],
        recommendedDependencies: [String] = [],
        kegOnly: Bool = false,
        revision: UInt32 = 0,
        versionScheme: UInt32 = 0
    ) {
        self.name = name
        self.fullName = fullName
        self.tap = tap
        self.desc = desc
        self.homepage = homepage
        self.license = license
        self.versions = versions
        self.bottle = bottle
        self.dependencies = dependencies
        self.buildDependencies = buildDependencies
        self.optionalDependencies = optionalDependencies
        self.testDependencies = testDependencies
        self.recommendedDependencies = recommendedDependencies
        self.kegOnly = kegOnly
        self.revision = revision
        self.versionScheme = versionScheme
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        name = try container.decode(String.self, forKey: .name)
        fullName = try container.decodeIfPresent(String.self, forKey: .fullName) ?? name
        tap = try container.decodeIfPresent(String.self, forKey: .tap) ?? "homebrew/core"
        desc = try container.decodeIfPresent(String.self, forKey: .desc)
        homepage = try container.decodeIfPresent(String.self, forKey: .homepage)
        license = try container.decodeIfPresent(String.self, forKey: .license)
        versions = try container.decode(FormulaVersions.self, forKey: .versions)
        bottle = try container.decodeIfPresent(BottleSpec.self, forKey: .bottle) ?? BottleSpec()
        dependencies = try container.decodeIfPresent([String].self, forKey: .dependencies) ?? []
        buildDependencies = try container.decodeIfPresent([String].self, forKey: .buildDependencies) ?? []
        optionalDependencies = try container.decodeIfPresent([String].self, forKey: .optionalDependencies) ?? []
        testDependencies = try container.decodeIfPresent([String].self, forKey: .testDependencies) ?? []
        recommendedDependencies = try container.decodeIfPresent([String].self, forKey: .recommendedDependencies) ?? []
        kegOnly = try container.decodeIfPresent(Bool.self, forKey: .kegOnly) ?? false
        revision = try container.decodeIfPresent(UInt32.self, forKey: .revision) ?? 0
        versionScheme = try container.decodeIfPresent(UInt32.self, forKey: .versionScheme) ?? 0
    }

    public var version: String {
        versions.stable
    }

    public var versionWithRevision: String {
        revision > 0 ? "\(versions.stable)_\(revision)" : versions.stable
    }
}

public struct FormulaVersions: Codable, Equatable, Sendable {
    public var stable: String
    public var head: String?
    public var bottle: Bool

    public init(stable: String, head: String? = nil, bottle: Bool = false) {
        self.stable = stable
        self.head = head
        self.bottle = bottle
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        stable = try container.decode(String.self, forKey: .stable)
        head = try container.decodeIfPresent(String.self, forKey: .head)
        bottle = try container.decodeIfPresent(Bool.self, forKey: .bottle) ?? false
    }
}

public struct BottleSpec: Codable, Equatable, Sendable {
    public var stable: BottleFiles?

    public init(stable: BottleFiles? = nil) {
        self.stable = stable
    }
}

public struct BottleFiles: Codable, Equatable, Sendable {
    public var rebuild: UInt32
    public var rootURL: String
    public var files: [String: BottleFile]

    enum CodingKeys: String, CodingKey {
        case rebuild
        case rootURL = "root_url"
        case files
    }

    public init(rebuild: UInt32 = 0, rootURL: String, files: [String: BottleFile]) {
        self.rebuild = rebuild
        self.rootURL = rootURL
        self.files = files
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        rebuild = try container.decodeIfPresent(UInt32.self, forKey: .rebuild) ?? 0
        rootURL = try container.decode(String.self, forKey: .rootURL)
        files = try container.decode([String: BottleFile].self, forKey: .files)
    }
}

public struct BottleFile: Codable, Equatable, Sendable {
    public var cellar: String
    public var url: String
    public var sha256: String

    public init(cellar: String, url: String, sha256: String) {
        self.cellar = cellar
        self.url = url
        self.sha256 = sha256
    }

    public var isRelocatable: Bool {
        cellar == ":any" || cellar == ":any_skip_relocation"
    }
}
